Add minimal unit tests for sailjaild

Tomin1 opened the pull request:
This implements a minimal set of unit tests that checks some of the
functionality in sailjaild. It should provide enough framework to expand
the tests later to cover all aspects of sailjaild. There is support for
running tests without installing and generating coverage reports for
ease of development.

Also a fix to util/strip function that was discovered to be a bit broken 
and also included support for building without libdbusaccess from #47.

There are convenience make commands to run tests locally without installing anything:
```bash
make -C daemon/test/ run       # Just build and run tests
make -C daemon/test/ coverage  # Build, run and create coverage reports
make -C daemon/test/ clean     # Clean up
```
However the main way to run these tests is automatic testing (see tests.xml.in).

LaakkonenJussi commented on daemon/test/Makefile:
```diff
@@ -0,0 +1,170 @@
+coverage:: coverage-run
+	lcov -c -d . -b $(CURDIR) -o $(PWD)/coverage
+	genhtml -o $(PWD)/report $(PWD)/coverage
+	xdg-open $(PWD)/report/index.html
```
Isn't this going a bit too far? Usually the path is only reported that can be then opened  by the user. Alternatively, an option for it could be that enables this  and by default it only reports?

Tomin1 replied:
Ah, this. It was convenient for me but yeah, I should have had echo or something instead. I will change it. :+1:

LaakkonenJussi commented on daemon/test/test_util.c:
```diff
@@ -0,0 +1,315 @@
+    keyfile_set_boolean(keyfile, "Foo", "boolean", true);
+    keyfile_set_integer(keyfile, "Foo", "number", 100);
+    g_assert_true(keyfile_save(keyfile, SHAREDSTATEDIR "/test.ini"));
+    g_assert_true(g_file_test(SHAREDSTATEDIR "/test.ini", G_FILE_TEST_EXISTS));
```
Should the content be verified as well?

Tomin1 replied:
Added that. Although I kind of expect that glib does its job when the file exists but I suppose it's still important to verify to some degree.

LaakkonenJussi commented:
Nothing special here, looks good except personally I'd like to see more negative tests with the minimal set. Now it seems that positive cases are mostly tested. Maybe add testing of interpretation invalid inputs to functions?

Tomin1 commented:
> Nothing special here, looks good except personally I'd like to see more negative tests with the minimal set. Now it seems that positive cases are mostly tested. Maybe add testing of interpretation invalid inputs to functions?

I added some of that, especially many invalid/null inputs. Also added branch coverage measurement with 'branches' target, which is the same as 'coverage' but also with branch coverage measurement (not all versions of lcov support that).

LaakkonenJussi commented:
Getting
```
ERROR:test_util.c:249:test_keyfile_merge: 'keyfile_merge(keyfile, TESTDATADIR "/keyfile1.ini")' should be TRUE
Bail out! ERROR:test_util.c:249:test_keyfile_merge: 'keyfile_merge(keyfile, TESTDATADIR "/keyfile1.ini")' should be TRUE
```
with daemon tests.

Looks legit, but couldn't yet run and check for coverage because of this.

LaakkonenJussi commented:
```
./test_appinfo; ./test_permissions; ./test_settings; ./test_stringset; ./test_util;
# random seed: R02S6c113db5ab03a5686a35260ae5af1fc3
1..4
# Start of sailjaild tests
# Start of appinfo tests
ok 1 /sailjaild/appinfo/create_missing
ok 2 /sailjaild/appinfo/create_invalid
**
ERROR:test_appinfo.c:109:test_appinfo_read_properties: 'appinfo_valid(appinfo)' should be TRUE
Bail out! ERROR:test_appinfo.c:109:test_appinfo_read_properties: 'appinfo_valid(appinfo)' should be TRUE
qemu: uncaught target signal 6 (Aborted) - core dumped
/bin/sh: line 1: 29115 Aborted                 ./test_appinfo
# random seed: R02Sc613f5051585626aad404bdc55110734
1..3
# Start of sailjaild tests
# Start of permissions tests
# GLib-GIO-DEBUG: _g_io_module_get_default: Found default implementation local (GLocalVfs) for ‘gio-vfs’
ok 1 /sailjaild/permissions/create_and_delete
**
ERROR:test_permissions.c:117:test_permissions_available: assertion failed (stringset_size(available) == 4): (0 == 4)
Bail out! ERROR:test_permissions.c:117:test_permissions_available: assertion failed (stringset_size(available) == 4): (0 == 4)
qemu: uncaught target signal 6 (Aborted) - core dumped
/bin/sh: line 1: 29117 Aborted                 ./test_permissions
# random seed: R02Sf04537ceea3cfaad70f3be74aca20de6
1..2
# Start of sailjaild tests
# Start of settings tests
# Start of settings tests
ok 1 /sailjaild/settings/settings/create_and_delete
**
ERROR:test_settings.c:183:test_settings_load: assertion failed (stringset_size(granted) == 1): (0 == 1)
Bail out! ERROR:test_settings.c:183:test_settings_load: assertion failed (stringset_size(granted) == 1): (0 == 1)
qemu: uncaught target signal 6 (Aborted) - core dumped
/bin/sh: line 1: 29120 Aborted                 ./test_settings

1..20
# Start of sailjaild tests
# Start of util tests
# Start of strip tests
ok 1 /sailjaild/util/strip/basic
ok 2 /sailjaild/util/strip/none
ok 3 /sailjaild/util/strip/null
# End of strip tests
# Start of path tests
ok 4 /sailjaild/util/path/basename
ok 5 /sailjaild/util/path/extension
ok 6 /sailjaild/util/path/dirname
ok 7 /sailjaild/util/path/to_desktop_name
ok 8 /sailjaild/util/path/from_permission_name
# End of path tests
# Start of change tests
ok 9 /sailjaild/util/change/uid
ok 10 /sailjaild/util/change/boolean
ok 11 /sailjaild/util/change/string
ok 12 /sailjaild/util/change/string_steal
# End of change tests
# Start of keyfile tests
**
ERROR:test_util.c:249:test_keyfile_merge: 'keyfile_merge(keyfile, TESTDATADIR "/keyfile1.ini")' should be TRUE
Bail out! ERROR:test_util.c:249:test_keyfile_merge: 'keyfile_merge(keyfile, TESTDATADIR "/keyfile1.ini")' should be TRUE
qemu: uncaught target signal 6 (Aborted) - core dumped
/bin/sh: line 1: 29124 Aborted                 ./test_util
```

These tests produce errors.

LaakkonenJussi commented:
> Getting
> 
> ```
> ERROR:test_util.c:249:test_keyfile_merge: 'keyfile_merge(keyfile, TESTDATADIR "/keyfile1.ini")' should be TRUE
> Bail out! ERROR:test_util.c:249:test_keyfile_merge: 'keyfile_merge(keyfile, TESTDATADIR "/keyfile1.ini")' should be TRUE
> ```
> 
> with daemon tests.
> 
> Looks legit, but couldn't yet run and check for coverage because of this.


> ```
 ...
> ```
> These tests produce errors.

Apparently these were the result of unclean build environment, ignore them.

But if I run first the tests with `make run` and after that `make coverage` I get:
```
Scanning . for .gcda files ...
geninfo: WARNING: no .gcda files found in . - skipping!
Finished .info-file creation
genhtml  -o /home/path/to/sailjail/sailjail/report /home/jlaakkonen/Code/projects/sailjail/sailjail/coverage
Reading data file /home/path/to/sailjail/sailjail/coverage
genhtml: ERROR: no valid records found in tracefile /home/path/to/sailjail/sailjail/coverage
make: *** [Makefile:86: coverage] Error 255
```
Running make clean in between does fix the issue and generate coverage reports. Should coverage then always run `clean`?

Tomin1 commented:
> But if I run first the tests with `make run` and after that `make coverage` I get:
> 
> ```
> Scanning . for .gcda files ...
> geninfo: WARNING: no .gcda files found in . - skipping!
> Finished .info-file creation
> genhtml  -o /home/path/to/sailjail/sailjail/report /home/jlaakkonen/Code/projects/sailjail/sailjail/coverage
> Reading data file /home/path/to/sailjail/sailjail/coverage
> genhtml: ERROR: no valid records found in tracefile /home/path/to/sailjail/sailjail/coverage
> make: *** [Makefile:86: coverage] Error 255
> ```
> 
> Running make clean in between does fix the issue and generate coverage reports. Should coverage then always run `clean`?

Yeah, that's because _run_ builds the tests without support for coverage and then if _coverage_ is run after that it won't rebuild the files because they are up to date and _make_ doesn't know that they had been built with wrong flags. Always running _clean_ beforehand would work I think but then running _coverage_ many times in a row would cause extra builds so I would rather not do that. I also don't want to include coverage flags by default as normally these are run in QA system where it's not needed or desired. That _coverage_ target is just a tool for those writing the tests.

If at some point we switch to a better build system, we could fix also this proper I think. Maybe I should add a comment about this bug for now.

LaakkonenJussi commented:
> > But if I run first the tests with `make run` and after that `make coverage` I get:
> > ```
> > Scanning . for .gcda files ...
> > geninfo: WARNING: no .gcda files found in . - skipping!
> > Finished .info-file creation
> > genhtml  -o /home/path/to/sailjail/sailjail/report /home/jlaakkonen/Code/projects/sailjail/sailjail/coverage
> > Reading data file /home/path/to/sailjail/sailjail/coverage
> > genhtml: ERROR: no valid records found in tracefile /home/path/to/sailjail/sailjail/coverage
> > make: *** [Makefile:86: coverage] Error 255
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Running make clean in between does fix the issue and generate coverage reports. Should coverage then always run `clean`?
> 
> Yeah, that's because _run_ builds the tests without support for coverage and then if _coverage_ is run after that it won't rebuild the files because they are up to date and _make_ doesn't know that they had been built with wrong flags. Always running _clean_ beforehand would work I think but then running _coverage_ many times in a row would cause extra builds so I would rather not do that. I also don't want to include coverage flags by default as normally these are run in QA system where it's not needed or desired. That _coverage_ target is just a tool for those writing the tests.
> 
> If at some point we switch to a better build system, we could fix also this proper I think. Maybe I should add a comment about this bug for now.

Fair enough. LGTM.

LaakkonenJussi reviewed (approved):
LGTM.

Tomin1 commented:
Rebased and added some more mockup functions because those were added to master branch.
